File: sdks/java/src/main/java/dev/puid/PuidError.java
package dev.puid;

/**
 * Thrown for any non-2xx API response and for client-side validation.
 *
 * <p>{@link #status()} is the HTTP status, or {@code null} for client-side / transport errors.
 * {@link #code()} is the API's machine-readable error code (e.g. {@code "rate_limited"}, {@code
 * "quota_exceeded"}, {@code "network_error"}), or {@code null}.
 */
public final class PuidError extends RuntimeException {

  private final Integer status;
  private final String code;

  public PuidError(String message, Integer status, String code) {
    super(message);
    this.status = status;
    this.code = code;
  }

  public PuidError(String message) {
    this(message, null, null);
  }

  /** HTTP status, or {@code null} for client-side / transport errors. */
  public Integer status() {
    return status;
  }

  /** Machine-readable error code, or {@code null}. */
  public String code() {
    return code;
  }
}

File: sdks/php/src/PuidError.php
<?php

declare(strict_types=1);

namespace Puid;

/**
 * Thrown for any non-2xx API response and for client-side validation.
 *
 * $status is the HTTP status (null for client-side / transport errors); $errorCode
 * is the API's machine-readable code (e.g. "rate_limited", "quota_exceeded").
 * (Named $errorCode rather than $code because \Exception already defines $code.)
 */
final class PuidError extends \RuntimeException
{
    public function __construct(
        string $message,
        public readonly ?int $status = null,
        public readonly ?string $errorCode = null,
    ) {
        parent::__construct($message);
    }
}
